src/dist_kmeans_init.h

```diff
@@ -197,6 +201,117 @@ namespace dist_custom {
+        kmi(Algo & algo, const daal::data_management::NumericTablePtr input)
+        {
+            int rank = MPI4DAAL::rank();
+            int nRanks = MPI4DAAL::nRanks();
+
+            // first determine total number of rows
+            size_t tot_rows = input->getNumberOfRows();
+            size_t start_row = tot_rows;
+            // first determine total number of rows
+            MPI4DAAL::allreduce(&tot_rows, 1, MPI_SUM);
+            // determine start of my chunk
+            MPI4DAAL::exscan(&start_row, 1, MPI_SUM);
+            if(rank==0) start_row = 0;
+
+            // Internal data to be stored on the local nodes
+            daal::data_management::DataCollectionPtr localNodeData;
+            // First step on each rank (output var will be used for output of step4 as well)
+            auto step14Out = algo.run_step1Local(input, tot_rows, start_row)->get(daal::algorithms::kmeans::init::partialCentroids);
+            // Only one rank actually computes centroids, we need to identify rank and bcast centroids to all others
+            int data_rank = not_empty(step14Out) ? rank : -1;
+            MPI4DAAL::allreduce(&data_rank, 1, MPI_MAX);
+            daal::data_management::NumericTablePtr step2In = MPI4DAAL::bcast(rank, nRanks, step14Out, data_rank);
+
+            // default value of nRounds used by all steps
+            const size_t nRounds = kmeans::init::Parameter(algo._nClusters).nRounds;
+
+            // vector with results of step2 for input into step5
+            std::vector<daal::data_management::NumericTablePtr> s2InForStep5;
+            if(rank == 0) {
+                s2InForStep5.push_back(step2In);
+            }
+
+            SharedPtr<daal::algorithms::kmeans::init::interface1::DistributedStep3MasterPlusPlusPartialResult> step3Output;
+
+            for(size_t iRound = 0; iRound < nRounds; ++iRound) {
+                // run step2 on each rank
+                auto s2res = algo.run_step2Local(input, localNodeData, step2In, false);
+                if(iRound==0) localNodeData = s2res->get(daal::algorithms::kmeans::init::internalResult);
+                auto s2Out = s2res->get(daal::algorithms::kmeans::init::outputOfStep2ForStep3);
+                //  and gather result on root
+                auto s3In = MPI4DAAL::gather(rank, nRanks, s2Out);
+
+                const int S34TAG = 3003;
+                // The input for s4 will be stored in s4In
+                daal::data_management::NumericTablePtr s4In;
+                // run step3 on root and send results to non-roots
+                if(rank == 0) {
+                    step3Output = algo.run_step3Master(s3In);
+                    s4In = step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, 0);
+                    for(int i=1; i<nRanks; i++) {
+                        MPI4DAAL::send(step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, i), i, S34TAG); // it can be NULL
```

Comment:
Why don't you follow what is done in plusplus method?
What does this change except that it makes it less scalable?